extensions/amp-form/0.1/amp-form.js

```diff
@@ -436,29 +448,75 @@ export class AmpForm {
    */
   handleXhrSubmit_(varSubsFields, trust) {
     this.setState_(FormState_.SUBMITTING);
-
-    const p = this.doVarSubs_(varSubsFields)
-        .then(() => {
-          this.triggerFormSubmitInAnalytics_('amp-form-submit');
-          this.actions_.trigger(
-              this.form_, 'submit', /* event */ null, trust);
-          // After variable substitution
-          const values = this.getFormAsObject_();
-          // At the form submitting state, we want to display any template
-          // messages with the submitting attribute.
-          this.renderTemplate_(values);
-        })
-        .then(() => this.doActionXhr_())
-        .then(response => this.handleXhrSubmitSuccess_(response),
-            error => {
-              return this.handleXhrSubmitFailure_(/** @type {!Error} */(error));
-            });
-
+    const varSubPromise = this.doVarSubs_(varSubsFields);
+    let p;
+    if (this.ssrTemplateHelper_.isSupported()) {
+      p = varSubPromise.then(() => {
+        this.actions_.trigger(
+            this.form_, 'submit', /* event */ null, trust);
+      }).then(() => {
+        return this.ssrTemplateHelper_.fetchAndRenderTemplate(
+            this.form_,
+            this.handleSsrTemplateSuccess_.bind(this),
+            this.handleSsrTemplateFailure_.bind(this)
+        );
+      });
+    } else {
+      p = varSubPromise
+          .then(() => {
+            this.handleFormSubmitting_(trust);
+          })
+          .then(() => this.doActionXhr_())
```

Comment:
Minor: `handleFormSubmitting_` is sync so we can combine and avoid an extra Promise tick.

```js
.then(() => {
  this.submittingWithTrust_(trust);
  return this.doActionXhr_();
})
```